Add support for critical_processes config option in ServerAgent #108

bodiakof opened the pull request:
### Summary:
- Introduced a new `critical_processes` attribute to `ServerAgent`, allowing specification of essential system processes.

bodiakof committed:
Added 'critical_processes=' field to config.ini file of web, dns, smtp and ntp

bodiakof committed:
Merge branch 'develop' into 53-detect-and-log-failed-processes-in-agent

bodiakof committed:
Add support for `critical_processes` config option in ServerAgent

- Introduced a new `critical_processes` attribute to `ServerAgent`, allowing specification of essential system processes.

- Added `self.critical_processes` to `__init__`, initialized as an empty list.
- Implemented a `@property` with setter for `critical_processes` to ensure only sequences (like lists or tuples) are assigned.
- Updated `_parse_config_file` to read `critical_processes` from the `[server]` section of the config file, if provided.

This feature allows the agent to track or respond to the presence/absence of specific critical processes. Enforcing the sequence type ensures robust error handling and prevents unexpected behavior due to improper configuration values.

bodiakof committed:
Add tests for critical_processes setter and config parsing

- Added a test to ensure assigning a non-sequence to critical_processes raises TypeError
- Extended config parsing tests to confirm comma-separated critical_processes are parsed as lists
- Added dedicated test to verify correct parsing of critical_processes from config file

bodiakof committed:
Fix UnboundLocalError in get_config_option by initializing alue to None when config.get fails

bodiakof committed:
Fix UnboundLocalError in get_config_option by initializing `value` to None

- Declare `value = None` before attempting config.get()
- Return `default` when `value` remains None instead of using `value or default`

VovaD199 commented on agents/dns/config.ini:
No values ​​filled in — all fields (except port) are empty. This makes the file non-functional without filling in.

No comments — it is not clear what is expected in each field.

No validation/types — it is not obvious what values ​​are allowed (comma-separated, URLs, ints, etc.).

Please contact the groups developing the relevant agents and try to fill in these fields, write docstrings.

bodiakof replied:
Config files are documented now, but most values gonna be added later, since it's not yet clear what's best to add. For example, 'interface=' field depends on OS, 'procesess=' supposed to be refactored. No need to fill 'name=' field since it's aleady assigned in agent's script.

VovaD199 commented on agents/utils/helpers.py:
```diff
@@ -41,6 +41,7 @@ def get_config_option(
         Note that by default, section names are case-sensitive, whereas
         option-names are case-insensitive.
     """
+    value = None
```
Please change the variable name from "value" to something more intuitive. Update this name in the function below.

VovaD199 commented on agents/tests/test_agent.py:
```diff
@@ -64,6 +69,45 @@ def test_type_checks_on_config_parse():
+
+        agent = MockAgent.from_config_file(tmp.name)
+        assert agent.critical_processes == ['sshd', 'nginx', 'postgres']
+
```
You can test an empty string or a value with spaces:
'critical_processes=   '

YelyzavetaV reviewed (commented):
While what you implemented is solid and works, I'm not happy with how critical_processes are implemented right now. Let me explain. We will definitely have some critical process that are unchangeably critical for ALL servers => we can implement them via global config. And then, if necessary, for each instance, append that list with more processes. This would be a bit similar to how `whitelist_commands` are now implemented - check it out and let me know what you think.

bodiakof committed:
Merge branch 'develop' into 53-detect-and-log-failed-processes-in-agent

bodiakof committed:
Changed variable name from 'value' to 'option_value' in agents/utils/configtools.py

bodiakof committed:
Merge branch 'develop' into 53-detect-and-log-failed-processes-in-agent

bodiakof committed:
Modified ntp.py, config.ini and congigtools.py files

- Added whitelist_commands, log_path to ntp.py and removed controller_url to be consistent with other scripts

- Added whitelist_command to config.ini

- Added 'option_value = None' at the start of get_config_option in configtools.py to avoid this error: If config.has_option(section, option) is False, then option_value never gets defined, leading to the UnboundLocalError.

bodiakof committed:
Merge branch 'develop' into 53-detect-and-log-failed-processes-in-agent

bodiakof committed:
Merge branch 'develop' into 53-detect-and-log-failed-processes-in-agent

bodiakof committed:
Add 'critical_processes' attribute to global.ini

bodiakof committed:
feat(agent): implement global + per-instance critical_processes

- Load always-critical processes from `global.ini` `[controller]critical_processes` into `GLOBAL_CRITICAL_PROCESSES`
- Seed `self.critical_processes` in `__init__` from global defaults on every instance
- Add `extra_critical_processes` parameter to `__init__` for code-level extensions
- Define `critical_processes` property with setter enforcing sequence type
- In `_parse_config_file`, read and append `[server]critical_processes` entries, avoiding duplicates

YelyzavetaV commented on agents/agent.py:
```diff
@@ -27,6 +28,16 @@
 
 PROTOCOLS = ('tcp', 'udp')
 
+# GLOBAL DEFAULTS FOR critical_processes
+config_path = os.path.join(os.path.dirname(__file__), 'global.ini')
+_global_cfg = ConfigParser.ConfigParser()
+_global_cfg.read(config_path)
+try:
+    raw = _global_cfg.get('controller', 'critical_processes')
+    GLOBAL_CRITICAL_PROCESSES = parse_csv_list(raw)
+except (ConfigParser.NoSectionError, ConfigParser.NoOptionError):
+    GLOBAL_CRITICAL_PROCESSES = []
+
```
Please see how `global.ini` is loaded in `develop` (in `agents/__init__.py`).

YelyzavetaV commented on agents/agent.py:
```diff
@@ -97,7 +108,9 @@ def __init__(
         interface=None,
         protocol=None,
         whitelist_commands=None,
+        extra_critical_processes=None,
```
We don't need two separate attributes: `critical_processes` and `extra_critical_processes`. Here is how it goes with whitelisted commands: all instances of agents have a class attribute upon instantiation `whitelist_commands`. Then, if additional commands are passed to an instance, a class attribute is overwritten and extended by an instance attribute with the same name. See the following code:

```python
if self.whitelist_commands is None:
    self.whitelist_commands = []

if whitelist_commands is not None:
    self.whitelist_commands.extend(whitelist_commands)
```

Or, when instance parameters are read from `config.ini`:

```python
whitelist_commands = get_config_option(
    config,
    'controller',
    'whitelist_commands',
    [],
    logger=self.logger,
    fallback_logger=self.fallback_logger,
    cast=parse_csv_list,
 )
 # Add commands to the list of globally allowed commands.
if whitelist_commands:
    self.whitelist_commands.extend(whitelist_commands)
```

YelyzavetaV commented on agents/agent.py:
```diff
@@ -113,6 +126,19 @@ def __init__(
+
+        if extra_critical_processes:
+            if not isinstance(extra_critical_processes, Sequence):
+                raise TypeError(
```
Type checks should be handled in properties.

YelyzavetaV commented on agents/agent.py:
```diff
@@ -113,6 +126,19 @@ def __init__(
+                    'critical_processes must be a sequence, not %s'
+                    % type(extra_critical_processes)
+                )
+            for proc in extra_critical_processes:
```
Use list comprehension for a more Python approach.

YelyzavetaV commented on agents/agent.py:
```diff
@@ -193,6 +219,19 @@ def processes(self, value):
+        return getattr(self, '_critical_processes', [])
+
+    @critical_processes.setter
+    def critical_processes(self, value):
```
You duplicate the type check here an in the constructor.

YelyzavetaV commented on agents/agent.py:
```diff
@@ -265,6 +304,33 @@ def _parse_config_file(self, filename=None):
                 self.whitelist_commands.extend(whitelist_commands)
 
+            # APPEND per-instance critical_processes
+            self.logger.debug(
+                'Looking up [server]/critical_processes in %r', filename
+            )
+            try:
+                raw = config.get('server', 'critical_processes')
+                self.logger.debug('Found raw critical_processes: %r', raw)
+                per_instance = parse_csv_list(raw)
+            except (
+                    ConfigParser.NoSectionError, ConfigParser.NoOptionError
+            ) as e:
+                self.logger.warning(
+                    'No per-instance critical_processes in %r: %s', filename, e
+                )
+                per_instance = []
+
+            for proc in per_instance:
+                if proc not in self.critical_processes:
+                    self.logger.debug(
+                        'Appending custom critical_process %r', proc
+                    )
+                    self.critical_processes.append(proc)
+            self.logger.info(
+                'After parsing, critical_processes = %r',
+                self.critical_processes
+            )
```
Please see how logging and reading parameters is done in the code above: we use helper functions `maybe_log_message` and `get_config_option`.

YelyzavetaV commented on agents/global.ini:
```diff
@@ -3,3 +3,4 @@ url=http://127.0.0.1:8000/
 api_prefix=api/v1/
 auth_token_type=
 whitelist_commands=uptime, df -h, ls, whoami
+critical_processes=ntpd, named, postfix, uvicorn
```
These are server-specific processes, they shouldn't be in `global.ini`.

YelyzavetaV commented on agents/ntp/config.ini:
```diff
@@ -1,8 +1,10 @@
 [server]
-name=
+name=ntp
```
This is already the default name assigned in `ntp.py`.

bodiakof committed:
Merge branch 'develop' into 53-detect-and-log-failed-processes-in-agent

bodiakof committed:
Modify global.ini, config.ini as requested

bodiakof committed:
Refactor loading of GLOBAL_CRITICAL_PROCESSES to use configtools.get_config_option with default

- Replaced manual ConfigParser loading and exception handling in agent.py
  with configtools.load_global_config and get_config_option for consistency
  with existing configuration loading in agents/__init__.py.
- Removed redundant exception handling by utilizing the 'default' parameter.

bodiakof committed:
refactor(agent): simplify critical_processes handling and config loading

- Removed redundant 'extra_critical_processes' parameter and related type checks.
- 'critical_processes' now loads defaults from GLOBAL_CRITICAL_PROCESSES only if unset.
- Extended 'critical_processes' list using config value from [server] section.
- Ensured no duplicate entries when merging config and defaults.

bodiakof committed:
fix(agent): always seed critical_processes from GLOBAL_CRITICAL_PROCESSES

- Remove incorrect None check on `self.critical_processes` (getter returns [])
- Directly initialize `self._critical_processes` with global defaults in __init__
- Ensures that agent instances always start with the global critical_processes list

bodiakof committed:
Merge branch 'develop' into 53-detect-and-log-failed-processes-in-agent

bodiakof committed:
refactor(agent): rename and centralize default critical_processes

- Renamed class-level `critical_processes` default to `default_critical_processes` to avoid name clash
- Updated `critical_processes` property getter to fall back on `default_critical_processes`
- Seeded instance `_critical_processes` from `default_critical_processes` in `__init__`
- Adjusted `agents/__init__.py` to set `ServerAgent.default_critical_processes` instead of `critical_processes`

YelyzavetaV committed:
refactor: minor refactor of critical processes

YelyzavetaV committed:
docs: document global config file and smtp config.ini

YelyzavetaV committed:
fix: fix ini comments

YelyzavetaV committed:
feat: add critical processes to concrete agents

bodiakof committed:
Merge branch 'develop' into 53-detect-and-log-failed-processes-in-agent

bodiakof committed:
Merge branch '53-detect-and-log-failed-processes-in-agent' of https://github.com/ITA-Dnipro/PyDataCenter5000 into 53-detect-and-log-failed-processes-in-agent

bodiakof committed:
docs: document config.ini of dns, ntp and web agents
